### PR TITLE
fix: fix race condition on datasource creation

### LINKED ISSUE #11572: [Bug]: Multiple queries on the same postgres datasource create new connection pools
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Suppose that there are 10 on page load actions of postgres type built on top of the same datasource. Also suppose the datasource has not created a connection pool so far (which would happen when no queries have run so far). When the said page is loaded, 10 queries are triggered in parallel. This would lead to all 10 queries creating a new connection pool for the datasource. This breaks the expectation that only one connection pool would be established per datasource. 

### Steps To Reproduce

1. Run appsmith on local and bring it up fresh. This ensures that we can reproduce this issue in the later steps.
2. On a new application, create 10 postgres queries on top of a single datasource. 
3. Go to settings page and set all these 10 queries to run on page load.
4. Now restart the appsmith container to ensure that the server comes up again and no connection pool exists to reproduce.
5. Now open the application created on step 2. 
6. Now observe that on the postgres database, 10 connections would exist instead <5.

### Environment

Production

### Version

Cloud

### LINKED ISSUE #14117: [Bug]: Postgresql connections are left idle.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

After running a query postgres connections are left idel

### Steps To Reproduce

1. Create Postgress data source.
2. Create multiple onload select queries.
3. Create multiple pages.
4. traverse multiple all the pages few times.
5. Check idle connections

### Public Sample App

https://release.app.appsmith.com/app/habitasaurus/page1-628f4177dfdca56c4283eb2a/edit/queries/628f4182dfdca56c4283eb2c

### Version

cloud

## Update:
- A large part of the issue is supposed to be resolved by this [fix](https://github.com/appsmithorg/appsmith/pull/14348). 
- Unsolved part is described [here](https://www.notion.so/appsmith/Postgres-multiple-idle-connections-issue-72cc080691c64385ab3efab5d573c630).

Front logo [Front conversations](https://app.frontapp.com/open/top_5sozw)

### PR DESCRIPTION
**Pull request by sumitsum**
## Description
- Currently when a datasource does not exist and multiple actions connected to the same datasource are run simultaneously then they all try to create a new connection object resulting in multiple connections being created to the datasource. 
- This PR attempts to solve the race condition problem defined above in the following way:  
  - enclose the datasource context creation logic within a critical section. 
  - the critical section above is synchronized based on datasource id. Hence, if two threads want to create a datasource on different datasource ids then they would not be synchronized. 
- Please note that the lock introduced in this PR will only work in a non clustered setup. To update the lock for a clustere setup, this [issue](https://github.com/appsmithorg/appsmith/issues/15265) has been created to track it.
- Added a new issue to track refactor change that was discussed as part of PR review: https://github.com/appsmithorg/appsmith/issues/14577 

Fixes #11572 #11202
Partially addresses #14117

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- JUnit TC
- Cypress
- Manual

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by nidhi-nair — APPROVED** [review]

**Comment by sumitsum** [comment]
As discussed with @prapullac, It is not possible to add cypress test for this.

**Review by nidhi-nair — APPROVED** [review]
LGTM

File: app/client/cypress/integration/Smoke_TestSuite/ServerSideTests/OnLoadTests/CheckPostgresConnOnPageLoad_spec.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      `select count(*) from pg_stat_activity where usename='test_conn_user_` +
+      guid +
+      `'`;
+    dataSources.EnterQuery(checkNoOfConnQuery);
+    cy.get(".CodeMirror textarea").focus();
+    dataSources.RunQuery();
+    dataSources.ReadQueryTableResponse(0).then(($cellData) => {
+      expect(Number($cellData)).to.lte(5);
+    });
+  });
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Smoke_TestSuite/ServerSideTests/OnLoadTests/CheckPostgresConnOnPageLoad_spec.ts:105]: Can we add case for clean up - Deleting both datasources & dropping user too!

**Review by Aishwarya-U-R — CHANGES_REQUESTED** [review]

File: app/client/cypress/integration/Smoke_TestSuite/ServerSideTests/OnLoadTests/CheckPostgresConnOnPageLoad_spec.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    cy.wrap("Postgres_2_ " + guid).as("dsName");
+    cy.get("@dsName").then(($dsName) => {
+      dsName = $dsName;
+    });
+  });
+
+  it("4. Create 10 queries", () => {
+    for (let i = 0; i < 10; i++) {
+      dataSources.NavigateFromActiveDS(dsName, true);
```
[the remark this answers is not in the document]
Comment (sumitsum) [reply app/client/cypress/integration/Smoke_TestSuite/ServerSideTests/OnLoadTests/CheckPostgresConnOnPageLoad_spec.ts]: As discussed, due to issue with cypress helper methods unfurling the datasource tab, we can't push the test case as it is right now. Hence, adding it to commented out section.

File: app/client/cypress/integration/Smoke_TestSuite/ServerSideTests/OnLoadTests/CheckPostgresConnOnPageLoad_spec.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    dataSources.StartDataSourceRoutes();
+  });
+
+  it("1. Create DS", () => {
+    agHelper.GenerateUUID();
+    cy.get("@guid").then((uid) => {
+      dataSources.NavigateToDSCreateNew();
+      dataSources.CreatePlugIn("PostgreSQL");
+      guid = uid;
+      agHelper.RenameWithInPane("Postgres " + guid, false);
```
[the remark this answers is not in the document]
Comment (sumitsum) [reply app/client/cypress/integration/Smoke_TestSuite/ServerSideTests/OnLoadTests/CheckPostgresConnOnPageLoad_spec.ts]: As discussed, due to issue with cypress helper methods unfurling the datasource tab, we can't push the test case as it is right now. Hence, adding it to commented out section.

File: app/client/cypress/integration/Smoke_TestSuite/ServerSideTests/OnLoadTests/CheckPostgresConnOnPageLoad_spec.ts (hunk 1 above)
Comment (sumitsum) [reply app/client/cypress/integration/Smoke_TestSuite/ServerSideTests/OnLoadTests/CheckPostgresConnOnPageLoad_spec.ts]: As discussed, due to issue with cypress helper methods unfurling the datasource tab, we can't push the test case as it is right now. Hence, adding it to commented out section.

**Review by Aishwarya-U-R — APPROVED** [review]
Approving cypress - analysis stage!

File: app/client/cypress/integration/Smoke_TestSuite/ServerSideTests/OnLoadTests/CheckPostgresConnOnPageLoad_spec.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      ee.DragDropWidgetNVerify("textwidget", i * 50 + 100, i * 50 + 250);
+      propPane.UpdatePropertyFieldValue("Text", "{{Query_" + i + ".data}}");
+      agHelper.ValidateNetworkStatus("@updateLayout", 200);
+    }
+    deployMode.DeployApp();
+    agHelper.Sleep(10000);
+    deployMode.NavigateBacktoEditor();
+  });
+
+  it("6. Run query to check number of open connections to Postgres db", () => {
```
[the remark this answers is not in the document]
Comment (sumitsum) [reply app/client/cypress/integration/Smoke_TestSuite/ServerSideTests/OnLoadTests/CheckPostgresConnOnPageLoad_spec.ts]: As discussed, due to issue with cypress helper methods unfurling the datasource tab, we can't push the test case as it is right now. Hence, adding it to commented out section.

**Review by nidhi-nair — APPROVED** [review]

**Pull request merged**